### PR TITLE
Add tests for language virtualization

### PR DESCRIPTION
**Pull request by olafurpg**
* Implemented all constructs from `LanguageVirtualization` except those
   mentioned in #27.
 * moved tests into separate files
 * improved debugging facilities with `debug` parameter to
   `DETransformer`

### CODE REVIEW

**Comment by olafurpg** [comment]
Review @vjovanov

File: dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala [hunk 1]
```diff
@@ -9,7 +9,7 @@ case object ValDef extends Exp[Int]
 case object NoArgs extends Exp[Int]
 case class JustTargs[T, U]() extends Exp[(T, U)]
 case class JustArgs(x: Exp[Int]) extends Exp[Int]
-case class ArgsAndTArgs[T, U](t: Exp[T], u: Exp[U]) extends Exp[(T, U)]
+case class ArgsAndTArgs[T, U](t: Exp[_], u: Exp[_]) extends Exp[(T, U)]
```
Comment (vjovanov) [inline dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala]: Why is this necessary?

**Comment by vjovanov** [comment]
Otherwise LGTM.

File: dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala (hunk 1 above)
Comment (olafurpg) [inline dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala]: It wasn't necessary, good catch!

I had to use existentials for the nested take and quickly assumed this one needed it too.

Comment (olafurpg) [inline dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala]: Fixed now.

File: directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala [hunk 2]
```diff
@@ -43,7 +43,7 @@ trait DslConfig {
    * @tparam T
    * @return Result type of the block
    */
-  def dsl[T](ast: Rep[T]): T
+  def dsl[T](ast: Rep[_]): T
```
Comment (vjovanov) [inline directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala]: Watch for the others. There is quite a few of them.

Comment (olafurpg) [inline directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala]: I thought that was necessary but now see I it wasn't. Will fix this.

Comment (vjovanov) [inline directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala]: There is quite a few others as well. Look through this commit.

File: dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala [hunk 3]
```diff
@@ -17,11 +17,35 @@ case class Take[T](self: Exp[_], n: Exp[Int]) extends Exp[TArgClassExample[T]]
...
+case class Infix_HashCode[T](t: Exp[T]) extends Exp[T]
+case class Infix_IsInstanceOf[T](t: Exp[_]) extends Exp[T]
+case class Infix_AsInstanceOf[T](t: Exp[_]) extends Exp[T]
+case class Infix_ToString[T](t: Exp[T]) extends Exp[T]
+case class Infix_Notify[T](t: Exp[T]) extends Exp[T]
+case class Infix_NotifyAll[T](t: Exp[T]) extends Exp[T]
+case class Infix_Wait[T](t: Exp[T]) extends Exp[T]
+case class Infix_WaitL[T](t: Exp[T], t1: Exp[_]) extends Exp[T]
+case class Infix_WaitTL[T](t: Exp[T], t1: Exp[_], l: Exp[_]) extends Exp[T]
+case class Infix_Synchronized[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
```
Comment (olafurpg) [inline dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala]: Synchronized will work like this
```scala
case class Infix_Synchronized[T](self: Exp[_], t1: Exp[T]) extends Exp[T]
```
but my gut says it should be more like this
```scala
case class Infix_Synchronized[T, U](self: Exp[T], t1: Exp[U]) extends Exp[U]
```
Should I leave the existential or is this a bug in the transformation? @vjovanov

File: dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala [hunk 4]
```diff
@@ -17,11 +17,35 @@ case class Take[T](self: Exp[_], n: Exp[Int]) extends Exp[TArgClassExample[T]]
...
+case class ReadVar[T](e: Exp[T]) extends Exp[T]
+case class Assign[T](x: Exp[T], t: Exp[_]) extends Exp[T]
+case class Infix_==[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
+case class Infix_Equals[T](t: Exp[T], t1: Exp[_]) extends Exp[T]
+case class Infix_Eq[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
+case class Infix_Ne[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
+case class Infix_!=[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
+case class Infix_##[T](t: Exp[T]) extends Exp[T]
+case class Infix_HashCode[T](t: Exp[T]) extends Exp[T]
+case class Infix_IsInstanceOf[T](t: Exp[_]) extends Exp[T]
```
Comment (olafurpg) [inline dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala]: The same issue as with `synchronized`. In my opinion, the transformation should provide the type of `t` as well as the type argument to `isInstanceOf[T]`. The signature of this class would then be something like this:

```scala
case class Infix_IsInstanceOf[T, U](t: Exp[T]) extends Exp[Boolean]
```
@vjovanov

**Comment by olafurpg** [comment]
I've pushed a fix for most of the type arguments, there is a question left in an outdated diffs whether there is a bug in the transformation.

File: dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala (hunk 3 above)
Comment (vjovanov) [inline dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala]: I think that the bug is in the definition of the case class only. The 
rest is fine.

File: dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala (hunk 4 above)
Comment (vjovanov) [inline dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala]: Yes, this should be boolean. The part after extends should always 
correspond to the result of the operation.

**Comment by vjovanov** [comment]
We can merge after the fixes. Great!

File: dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala (hunk 4 above)
Comment (olafurpg) [inline dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala]: It is already fixed then, see [here](https://github.com/directembedding/directembedding/pull/28/files#diff-89f46380466422ec546cad3146575ce6R37)

Comment (vjovanov) [inline dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala]: Can I merge?

Comment (olafurpg) [inline dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala]: Yes, you can merge. I was able to remove all existential types except four which are all the same case as this one.
